docs/source/cpp/gandiva.rst

```diff
@@ -0,0 +1,151 @@
+or :class:`gandiva::Condition`, depending on how they will be used.
+``Expression`` is used in projections while ``Condition`` is used filters.
+
+As an example, here is how to create an Expression representing ``x + 3`` and a
+Condition representing ``x < 3``:
+
+.. code-block:: cpp
+
+   auto field_x_raw = arrow::field("x", arrow::int32());
+   auto field_x = TreeExprBuilder::MakeField(field_x_raw);
+   auto literal_3 = TreeExprBuilder::MakeLiteral(3);
+   auto field_result = arrow::field("result", arrow::int32());
+
+   auto add_node = TreeExprBuilder::MakeFunction("add", {field_x, literal_3}, arrow::int32());
+   auto expression = TreeExprBuilder::MakeExpression(add_node, field_result);
+
+   auto less_than_node = TreeExprBuilder::MakeFunction("less_than", {field_x, literal_3},
+                                                       boolean());
+   auto condition = TreeExprBuilder::MakeCondition(less_than_node);
+
+For simpler expressions, there are also convenience functions that allow you to
+use functions directly in ``MakeExpression`` and ``MakeCondition``:
+
+.. code-block:: cpp
+
+   auto expression = TreeExprBuilder::MakeExpression("add", {field_x, literal_3}, field_result);
+
+   auto condition = TreeExprBuilder::MakeCondition("less_than", {field_x, literal_3});
+
+
+Projectors and Filters
+======================
+
+Gandiva's two execution kernels are :class:`gandiva::Projector` and
+:class:`gandiva::Filter`. ``Projector`` consumes a record batch and projects
+into a new record batch. ``Filter`` consumes a record batch and produces a
+:class:`gandiva::SelectionVector` containing the indices that matched the condition.
+
+For both ``Projector`` and ``Filter``, optimization of the expression IR happens
+when creating instances. They are compiled against a static schema, so the
+schema of the record batches must be known at this point.
+
+Continuing with the ``expression`` and ``condition`` created in the previous
+section, here is an example of creating a Projector and a Filter:
+
+.. code-block:: cpp
+
+   auto schema = arrow::schema({field_x});
+   std::shared_ptr<Projector> projector;
+   auto status = Projector::Make(schema, {expression}, &projector);
+   ARROW_CHECK_OK(status);
+
+   std::shared_ptr<Filter> filter;
+   status = Filter::Make(schema, condition, &filter);
+   ARROW_CHECK_OK(status);
+
+
+Once a Projector or Filter is created, it can be evaluated on Arrow record batches.
+These execution kernels are single-threaded on their own, but are designed to be
+reused to process distinct record batches in parallel.
+
+Execution is performed with :func:`gandiva::Projector::Evaluate` and
+:func:`gandiva::Filter::Evaluate`. Filters produce :class:`gandiva::SelectionVector`,
+a vector of row indices that matched the filter condition. When filtering and
+projecting record batches, you can pass the selection vector into the projector
+so that the projection is only evaluated on matching rows.
```

Comment:
So you can pass the `SelectionVector` into `Projector::Evaluate()`? If so, it may be worth calling those out explicitly.

Reply:
> So you can pass the SelectionVector into Projector::Evaluate()?

Yes, there is an overload for that. It's documented in the API reference.

I'm not sure I understand your suggestion. Are you saying something like this?

```
For filtering and
projecting record batches simultaneously, you can pass the selection vector into the 
projector (using one of the overloads) so that the projection is only evaluated on 
matching rows.
```

Reply:
I mean, instead of saying "pass the selection vector into the projector," say "pass this `SelectionVector` into `Projector::Evaluate()`. It's subtle, but gives people access to links and makes it very clear how to move things around.